front-end/src/visualization/Visualization.jsx

```diff
@@ -50,47 +57,115 @@ function Visualization() {
-            const songId = clusterId * 100 + index + 1;
-            nodes.push({ id: songId, label: `Song ${index + 1}`, x: pos.x, y: pos.y, color: clusterColor });
-            edges.push({ from: clusterId, to: songId, color: 'white' });
+        clusterData.centroids.forEach((centroid, index) => {
+            const formattedCentroid = clusterData.centroids[index]?.toFixed(2) || '0.00';
+            const clusterLabel = `Valence: ${formattedCentroid}`;
+            const { x, y } = clusterPositions[index];
+            const songsInCluster = clusterData.assignments.filter(song => song.cluster === index);
+            const songPositions = calculateCircularPositions(x, y, 100, songsInCluster.length);
+            const clusterColor = getRandomColor();
+
+            const nodes = [{ id: globalId++, label: clusterLabel, x, y, isCluster: true, color: clusterColor }];
+            const edges = [];
+
+            songsInCluster.forEach((song, idx) => {
+                nodes.push({
+                    id: song.id,
+                    label: song.name,
+                    x: songPositions[idx].x,
+                    y: songPositions[idx].y,
+                    color: clusterColor,
+                    imageUrl: song.album.images[0].url
+                });
+                edges.push({ from: globalId - 1, to: song.id, color: 'white' });
+            });
+
+            allNodes = allNodes.concat(nodes);
+            allEdges = allEdges.concat(edges);
         });
 
-        return { nodes, edges };
+        return { nodes: allNodes, edges: allEdges };
     }
 
-    function addInterClusterEdges(nodes, edges) {
-        const song1 = nodes.find(node => node.label === 'Song 1' && node.id > 100 && node.id < 200);
-        const song3 = nodes.find(node => node.label === 'Song 3' && node.id > 200 && node.id < 300);
-        if (song1 && song3) {
-            edges.push({
-                from: song1.id,
-                to: song3.id,
-                color: 'red'
-            });
+    useEffect(() => {
+        const getClusterData = async () => {
+            try {
+                const response = await fetch(`${import.meta.env.VITE_BACKEND_ADDRESS}/clusters/${stitchId}`);
+                const data = await response.json();
+                setClusterData(data);
+            } catch (error) {
+                console.error('Error getting clusters:', error);
+            }
+        };
+
+        const fetchStitchName = async () => {
+            try {
+                const url = `${import.meta.env.VITE_BACKEND_ADDRESS}/stitch/title/${stitchId}`;
+                const response = await fetch(url);
+
+                if (!response.ok) {
+                    throw new Error('Failed to fetch stitch name');
+                }
+
+                const data = await response.json();
+                setTitle(data.title || 'Untitled');
+            } catch (error) {
+                console.error('Error fetching stitch name:', error);
```

Comment:
Same comment as above. While error handling is not a P0 for this project, filling the console with errors is definitely not desirable.